File: content/problems/1883-minimum-skips-to-arrive-at-meeting-on-time/solutions/scaled-time-skip-dp/Solution.java
class Solution {
    public int minSkips(int[] dist,int speed,int hoursBefore){
        long infinity=Long.MAX_VALUE/4;
        long[] dp=new long[dist.length+1];
        Arrays.fill(dp,infinity);
        dp[0]=0;
        for(int road=0;road<dist.length;road++){
            long[] next=new long[dist.length+1];
            Arrays.fill(next,infinity);
            for(int skipped=0;skipped<=road;skipped++)if(dp[skipped]<infinity){
                long arrival=dp[skipped]+dist[road];
                long waited=road==dist.length-1?arrival:(arrival+speed-1)/speed*speed;
                next[skipped]=Math.min(next[skipped],waited);
                if(road<dist.length-1)next[skipped+1]=Math.min(next[skipped+1],arrival);
            }
            dp=next;
        }
        long limit=(long)hoursBefore*speed;
        for(int skipped=0;skipped<=dist.length;skipped++)if(dp[skipped]<=limit)return skipped;
        return -1;
    }
}

File: content/problems/1883-minimum-skips-to-arrive-at-meeting-on-time/solutions/scaled-time-skip-dp/Solution.kt
class Solution {
    fun minSkips(dist:IntArray,speed:Int,hoursBefore:Int):Int{
        val infinity=Long.MAX_VALUE/4;
        var dp=LongArray(dist.size+1){
            infinity
        };
        dp[0]=0;
        for(road in dist.indices){
            val next=LongArray(dist.size+1){
                infinity
            };
            for(skipped in 0..road)if(dp[skipped]<infinity){
                val arrival=dp[skipped]+dist[road];
                val waited=if(road==dist.lastIndex)arrival else (arrival+speed-1)/speed*speed;
                next[skipped]=minOf(next[skipped],waited);
                if(road<dist.lastIndex)next[skipped+1]=minOf(next[skipped+1],arrival)
            };
            dp=next
        };
        val limit=hoursBefore.toLong()*speed;
        for(skipped in dp.indices)if(dp[skipped]<=limit)return skipped;
        return -1
    }
}
